### ISSUE TITLE
Do not use wall-clock time to handle delays

### ISSUE THREAD

**Issue by ghost** [input-output-hk/ce-netsim#30]
Netsim currently handles delays and more generally time using wall-clock, e.g messages are delivered after some delay configuration in the simulation based on [real-time clock](https://github.com/input-output-hk/ce-netsim/blob/31d43b4cc10e0bbab054ed936ab5f926f22f64b8/netsim-core/src/sim_context.rs#L157). 

In the context of a network simulation, this is problematic as it makes it impractical to test large scale simulations with realistic timings from a global network. For example, we know that the average delay for block propagation in Cardano is around 1s, if we want to simulate the effect of more or less delays and the impact of bandwidth consumption caused by larger blocks or more traffic in the network (eg. Peras votes), we would need to configure appropriately the latency of various links in our `netsim` network which would lead to unbearably slow simulations.

What's needed is a way to advance time much faster, like it's suggested in [this paper](https://www.informs-sim.org/wsc11papers/291.pdf): Roughly, time is governed by a control thread which allows worker threads to advance until some point. Latency is simulated and messages delivered only when simulation time has advanced far enough.

Ping @bwbush

**Comment by NicolasDP** [comment]
In the end you are still limited by the computing power of your machine and the workers still share that same computing power (so far, we can still imagine a distributed network simulator). Going faster will make things even more difficult for the agents to complete on time.

**Comment by abailly-iohk** [comment]
That's true only if your simulation does the same work than the actual implementation, which makes your simulation not simulate much :) For example we know that a great deal of time in Cardano is spent validating crypto stuff (keys, VRFs proofs) and then verifying the ledger state (validating txs, managing UTxOs, smart contracts...). This is irrelevant for a consensus protocol: the crypto can be completely faked, or itself simulated to represent special actions from the adversary, and whether or not a block is valid can also be simulated without the need to implement actual rules.

**Comment by NicolasDP** [comment]
in that case you can also configure your consensus to produce blocks at an higher rate than the real implementation (for example you can have 1 slot every 10ms and an avg of a block every 200ms) and accelerate the protocol testing 100x. We currently have an average of 11.787µs derivation from the expected latency (this is our overhead). So you can easily explore the sub seconds world in your consensus configuration to test a protocol.

**Comment by abailly-iohk** [comment]
Sure, I could do that. But this means change in the "production" code which might have unwanted effects. I would rather have the simulation environment take care of time

**Comment by SupernaviX** [comment]
@NicolasDP I'm using this library for the Leios sim, and the dependency on wall-clock time is making it difficult to work with. The simulation is nondeterministic (messages sent to different nodes within an `idle_duration` span can be seen in any order), and it's coupled to real time (at best I can make it run at 10x or 100x real time, but the more things that happen within `idle_duration`, the more the traffic gets jumbled). The rest of the simulation can use a simple event loop and run as fast as the CPU can handle it, but netsim means I have to use sleeps everywhere anyway.

It would be far more useful to me to have access to `SimMuxCore` directly. That component has no hard dependency on real time (it uses `Instant` but never `Instant::now()`), so I could "advance" time in the event loop. It would also remove the need for `idle_duration` since there wouldn't be a thread looping in the background, so no more nondeterminism. Do you think exposing that would make sense?

**Comment by NicolasDP** [comment]
`netsim` will add latency and bandwidth to the messages that are sent through the simulated network. This means that they **will** get "jumbled" and they will not be delivered directly as soon as sent. The goal of `netsim` is to provide a close to reality feel of network communication without a complicated user experience. `idle_duration` may not be an accurate name for what this is doing now. It is more a refresh speed or a granularity cursor. You should feel free to lower the number to a much lower value (like `10μs`) depending on the size of your simulation and the amount of data you are pushing through the network. Current default is `500μs` and this is usually good enough for sub millisecond precision of the simulation -- A local ping from my computer to the router will be usually within 1 and 2ms.

Depending on which `netsim` or `netsim-async` you are using you can avoid pausing your threads by using the blocking API so that you do not have to await when a message arrives but react to events. There are a few examples in the repository.

Now as to changing the Wall Clock dependency that is the core issue of this Github Issue. I think this is the first time I'm hearing the requirements of running simulation 10x or 100x real time. And I understand how this can be useful. Same for pausing and restarting the simulation. I see value in doing this and I can imagine how this could be done without too much fuss in the code. I already made some changes by removing the dependency on `SystemTime` which was initially used for the MVP. We are now using `Instant` which reflects a bit more of an abstract time progression. As you can see the calls to `Instant::now` are fairly isolated. If this helps I'd be happy to add command instruction to the message bus to: Start, Pause, Accelerate or Slowdown. Which would:

* start the internal clock,
* pause the internal clock,
* accelerate the progression of the internal clock,
* slow down the progression of the internal clock,

Would that be helpful and would this solve the problem ?

**Comment by SupernaviX** [comment]
> netsim will add latency and bandwidth to the messages that are sent through the simulated network. This means that they will get "jumbled" and they will not be delivered directly as soon as sent.

I understand this, simulating a delay for message transmission is why we want to use `netsim`. I don't expect messages to arrive in the order they were sent. When I say messages are "jumbled", I'm referring to messages arriving in a different order than they should have been _received_. I understand that `idle_duration` is configurable, and can be reduced (at the cost of eating more CPU time). I'm looking for a solution to this that doesn't involve a polling background thread, and so doesn't have to make that tradeoff. We're planning to simulate networks of a few thousand nodes with significantly more traffic than today, and I want to make sure the solution scales.

> Depending on which netsim or netsim-async you are using you can avoid pausing your threads by using the blocking API so that you do not have to await when a message arrives but react to events. There are a few examples in the repository.

I don't think a blocking API solves my problem. It still has the problem of the simulation staying coupled to realtime. Did you have a specific example in mind?

As for "Start, Pause, Accelerate, or Slowdown", I don't think that would be helpful for me. I'm already using an approach similar to what you described above:
- The simulation is sped up by a multiplier I called "timescale", such as 100x realtime.
- All network latencies are divided by that "timescale"; if a link is supposed to have `33ms` latency, it's configured with `0.33ms` instead.
- The simulation tracks all events using a `Timestamp`, representing the duration since the sim began. Converting between `Instant` and `Timestamp` involves dividing by "timescale". If a message arrives `0.33ms` after the sim began, we treat its arrival timestamp is `33ms`.

Accelerating the netsim clock wouldn't be useful, because the sim has scheduled events (such as "new slot" every "1 second"). Because netsim messages arrive in realtime, I need to add sleeps when waiting for those events:
```rust
// Return the next event that the simulation must process.
// I would like to implement this method without sleeps and without blocking the thread.
async fn next_event() -> Event {
    let next_queued_event = async move {
        let (event, timestamp) = compute_next_scheduled_event();
        // The sleep below is the problem. It's necessary because netsim has its own clock,
        // and the Leios simulation needs to stay in sync with the netsim clock.
        sleep_until(timestamp).await;
        (event, timestamp)
    };

    return select! {
        event = next_queued_event => event
        msg = get_next_incoming_netsim_message() => Event::IncomingMessage(msg)
    };
}
```

If the sim has an event scheduled `1000ms` from now, and netsim is running at 100x realtime, the sim needs to sleep for `10ms` before handling that event. "Accelerate" and "Slowdown" don't really help me here, because that sleep is outside of netsim's code. My code still has to keep its own clock in sync with netsim's, and speeding up/slowing down netsim's clock just makes that more complicated.

I'm looking for a solution where netsim doesn't have an internal clock, and isn't "in charge" of the passage of time. Then the sim could run as quickly as possible, without loss of accuracy from `idle_duration`, and probably with less CPU usage because it wouldn't need a background thread waking up every 500μs or less.
